File: frontend/android-app/app/src/main/java/com/codenized/planixor/domain/backup/BackupDeserializer.kt
package com.codenized.planixor.domain.backup

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import javax.inject.Inject

/**
 * Deserializes a backup JSON string into a BackupFile domain model.
 * Configured to ignore unknown JSON fields for forward compatibility.
 *
 * Supports both the current named-key format and the legacy minified format
 * (single-letter keys: a, b, c, etc.) produced by earlier versions of the app.
 */
class BackupDeserializer @Inject constructor() {

    private val gson: Gson = GsonBuilder()
        .create()

    /**
     * Parses a JSON string into a BackupFile structure.
     * If the legacy minified format is detected (root has "a" and "b" keys instead
     * of "metadata" and "data"), it's automatically converted to the current format.
     * Unknown fields at any nesting level are silently ignored.
     *
     * @param json The raw JSON string from the backup file
     * @return Parsed BackupFile with all recognized fields populated
     * @throws com.google.gson.JsonSyntaxException if the JSON is malformed
     */
    fun deserialize(json: String): BackupFile {
        val root = JsonParser.parseString(json).asJsonObject

        // Detect legacy minified format (has "a" and "b" keys, no "metadata" key)
        if (root.has("a") && root.has("b") && !root.has("metadata")) {
            val converted = convertLegacyFormat(root)
            return gson.fromJson(converted, BackupFile::class.java)
        }

        return gson.fromJson(json, BackupFile::class.java)
    }

    /**
     * Converts legacy minified format to the current named-key format.
     *
     * Legacy key mapping:
     * - Root: a=metadata, b=data
     * - Metadata: a=createdAt
     * - Data: a=calendarEvents, b=notificationRecords, c=annualHoursConfig, d=shifts, e=reminders
     * - CalendarEvent: a=id, b=eventType, c=eventTypeId, d=startDay, e=endDay, f=startTime,
     *   g=endTime, h=totalHours, i=notes, j=alertOffsets, k=modifiedAt, l=syncedAt, m=isDeleted
     * - NotificationRecord: a=id, b=calendarEventId, c=alertOffset, d=triggerTime,
     *   e=isDelivered, f=isRead, g=modifiedAt, h=syncedAt, i=isDeleted
     * - AnnualHoursConfig: a=id, b=year, c=configuredHours, d=modifiedAt, e=syncedAt, f=isDeleted
     * - Shift: a=id, b=name, c=icon, d=backgroundColor, e=startTime, f=endTime,
     *   g=hoursWorked, h=isActive, i=createdAt, j=modifiedAt, k=syncedAt, l=isDeleted
     * - Reminder: a=id, b=name, c=icon, d=backgroundColor, e=isActive,
     *   f=createdAt, g=modifiedAt, h=syncedAt, i=isDeleted
     */
    private fun convertLegacyFormat(root: JsonObject): JsonObject {
        val legacyMeta = root.getAsJsonObject("a") ?: JsonObject()
        val legacyData = root.getAsJsonObject("b") ?: JsonObject()

        val result = JsonObject()

        // Metadata
        val metadata = JsonObject()
        metadata.addProperty("createdAt", legacyMeta.get("a")?.asString ?: "")
        metadata.addProperty("appVersion", "1.0.0")
        metadata.addProperty("platform", "web")
        metadata.addProperty("schemaVersion", 1)
        result.add("metadata", metadata)

        // Data
        val data = JsonObject()

        // Calendar events (data.a)
        val events = com.google.gson.JsonArray()
        legacyData.getAsJsonArray("a")?.forEach { element ->
            val e = element.asJsonObject
            val event = JsonObject()
            event.addProperty("id", e.get("a")?.asString ?: "")
            event.addProperty("eventType", e.get("b")?.asString ?: "")
            event.addProperty("eventTypeId", e.get("c")?.asString ?: "")
            event.addProperty("startDay", e.get("d")?.asString ?: "")
            event.addProperty("endDay", e.get("e")?.asString ?: "")
            event.addProperty("startTime", e.get("f")?.asInt ?: 0)
            event.addProperty("endTime", e.get("g")?.asInt ?: 0)
            event.addProperty("totalHours", e.get("h")?.asInt ?: 0)
            if (e.get("i")?.isJsonNull != false) event.add("notes", com.google.gson.JsonNull.INSTANCE)
            else event.addProperty("notes", e.get("i")?.asString)
            event.add("alertOffsets", e.getAsJsonArray("j") ?: com.google.gson.JsonArray())
            event.addProperty("modifiedAt", e.get("k")?.asString ?: "")
            if (e.get("l")?.isJsonNull != false) event.add("syncedAt", com.google.gson.JsonNull.INSTANCE)
            else event.addProperty("syncedAt", e.get("l")?.asString)
            event.addProperty("isDeleted", e.get("m")?.asBoolean ?: false)
            events.add(event)
        }
        data.add("calendarEvents", events)

        // Notification records (data.b)
        val notifications = com.google.gson.JsonArray()
        legacyData.getAsJsonArray("b")?.forEach { element ->
            val e = element.asJsonObject
            val record = JsonObject()
            record.addProperty("id", e.get("a")?.asString ?: "")
            record.addProperty("calendarEventId", e.get("b")?.asString ?: "")
            record.addProperty("alertOffset", e.get("c")?.asInt ?: 0)
            record.addProperty("triggerTime", e.get("d")?.asString ?: "")
            record.addProperty("isDelivered", e.get("e")?.asBoolean ?: false)
            record.addProperty("isRead", e.get("f")?.asBoolean ?: false)
            record.addProperty("modifiedAt", e.get("g")?.asString ?: "")
            if (e.get("h")?.isJsonNull != false) record.add("syncedAt", com.google.gson.JsonNull.INSTANCE)
            else record.addProperty("syncedAt", e.get("h")?.asString)
            record.addProperty("isDeleted", e.get("i")?.asBoolean ?: false)
            notifications.add(record)
        }
        data.add("notificationRecords", notifications)

        // Annual hours config (data.c)
        val annualHours = com.google.gson.JsonArray()
        legacyData.getAsJsonArray("c")?.forEach { element ->
            val e = element.asJsonObject
            val config = JsonObject()
            config.addProperty("id", e.get("a")?.asString ?: "")
            config.addProperty("year", e.get("b")?.asInt ?: 0)
            config.addProperty("configuredHours", e.get("c")?.asInt ?: 0)
            config.addProperty("modifiedAt", e.get("d")?.asString ?: "")
            if (e.get("e")?.isJsonNull != false) config.add("syncedAt", com.google.gson.JsonNull.INSTANCE)
            else config.addProperty("syncedAt", e.get("e")?.asString)
            config.addProperty("isDeleted", e.get("f")?.asBoolean ?: false)
            annualHours.add(config)
        }
        data.add("annualHoursConfig", annualHours)

        // Shifts (data.d)
        val shifts = com.google.gson.JsonArray()
        legacyData.getAsJsonArray("d")?.forEach { element ->
            val e = element.asJsonObject
            val shift = JsonObject()
            shift.addProperty("id", e.get("a")?.asString ?: "")
            shift.addProperty("name", e.get("b")?.asString ?: "")
            shift.addProperty("icon", e.get("c")?.asString ?: "")
            shift.addProperty("backgroundColor", e.get("d")?.asString ?: "")
            shift.addProperty("startTime", e.get("e")?.asInt ?: 0)
            shift.addProperty("endTime", e.get("f")?.asInt ?: 0)
            shift.addProperty("hoursWorked", e.get("g")?.asInt ?: 0)
            shift.addProperty("isActive", e.get("h")?.asBoolean ?: true)
            shift.addProperty("createdAt", e.get("i")?.asString ?: "")
            shift.addProperty("modifiedAt", e.get("j")?.asString ?: "")
            if (e.get("k")?.isJsonNull != false) shift.add("syncedAt", com.google.gson.JsonNull.INSTANCE)
            else shift.addProperty("syncedAt", e.get("k")?.asString)
            shift.addProperty("isDeleted", e.get("l")?.asBoolean ?: false)
            shifts.add(shift)
        }
        data.add("shifts", shifts)

        // Reminders (data.e)
        val reminders = com.google.gson.JsonArray()
        legacyData.getAsJsonArray("e")?.forEach { element ->
            val e = element.asJsonObject
            val reminder = JsonObject()
            reminder.addProperty("id", e.get("a")?.asString ?: "")
            reminder.addProperty("name", e.get("b")?.asString ?: "")
            reminder.addProperty("icon", e.get("c")?.asString ?: "")
            reminder.addProperty("backgroundColor", e.get("d")?.asString ?: "")
            reminder.addProperty("isActive", e.get("e")?.asBoolean ?: true)
            reminder.addProperty("createdAt", e.get("f")?.asString ?: "")
            reminder.addProperty("modifiedAt", e.get("g")?.asString ?: "")
            if (e.get("h")?.isJsonNull != false) reminder.add("syncedAt", com.google.gson.JsonNull.INSTANCE)
            else reminder.addProperty("syncedAt", e.get("h")?.asString)
            reminder.addProperty("isDeleted", e.get("i")?.asBoolean ?: false)
            reminders.add(reminder)
        }
        data.add("reminders", reminders)

        // Sync config (data.f) — pass through as empty if not present
        data.add("syncConfig", legacyData.getAsJsonArray("f") ?: com.google.gson.JsonArray())

        result.add("data", data)
        return result
    }
}

File: frontend/react-web/src/features/backup/services/backupDeserializer.ts
import type {
  BackupAnnualHoursConfig,
  BackupCalendarEvent,
  BackupData,
  BackupFile,
  BackupMetadata,
  BackupNotificationRecord,
  BackupReminder,
  BackupShift,
  BackupSyncConfig,
} from '../models';

/**
 * Detects whether the parsed JSON uses the legacy minified key format.
 * Legacy format uses single-letter keys: root has "a" (metadata) and "b" (data).
 */
const isLegacyFormat = (parsed: Record<string, unknown>): boolean => {
  return 'a' in parsed && 'b' in parsed && !('metadata' in parsed);
};

/**
 * Converts a legacy minified backup to the current format structure.
 *
 * Legacy key mapping:
 * - Root: a=metadata, b=data
 * - Metadata: a=createdAt
 * - Data: a=calendarEvents, b=notificationRecords, c=annualHoursConfig, d=shifts, e=reminders, f=syncConfig
 * - CalendarEvent: a=id, b=eventType, c=eventTypeId, d=startDay, e=endDay, f=startTime, g=endTime, h=totalHours, i=notes, j=alertOffsets, k=modifiedAt, l=syncedAt, m=isDeleted
 * - NotificationRecord: a=id, b=calendarEventId, c=alertOffset, d=triggerTime, e=isDelivered, f=isRead, g=modifiedAt, h=syncedAt, i=isDeleted
 * - AnnualHoursConfig: a=id, b=year, c=configuredHours, d=modifiedAt, e=syncedAt, f=isDeleted
 * - Shift: a=id, b=name, c=icon, d=backgroundColor, e=startTime, f=endTime, g=hoursWorked, h=isActive, i=createdAt, j=modifiedAt, k=syncedAt, l=isDeleted
 * - Reminder: a=id, b=name, c=icon, d=backgroundColor, e=isActive, f=createdAt, g=modifiedAt, h=syncedAt, i=isDeleted
 */
const convertLegacyFormat = (parsed: Record<string, unknown>): Record<string, unknown> => {
  const legacyMeta = (parsed.a ?? {}) as Record<string, unknown>;
  const legacyData = (parsed.b ?? {}) as Record<string, unknown>;

  const calendarEvents = Array.isArray(legacyData.a)
    ? (legacyData.a as Record<string, unknown>[]).map((e) => ({
        id: e.a, eventType: e.b, eventTypeId: e.c, startDay: e.d, endDay: e.e,
        startTime: e.f, endTime: e.g, totalHours: e.h, notes: e.i,
        alertOffsets: e.j, seriesId: null, modifiedAt: e.k, syncedAt: e.l, isDeleted: e.m,
      }))
    : [];

  const notificationRecords = Array.isArray(legacyData.b)
    ? (legacyData.b as Record<string, unknown>[]).map((e) => ({
        id: e.a, calendarEventId: e.b, alertOffset: e.c, triggerTime: e.d,
        isDelivered: e.e, isRead: e.f, modifiedAt: e.g, syncedAt: e.h, isDeleted: e.i,
      }))
    : [];

  const annualHoursConfig = Array.isArray(legacyData.c)
    ? (legacyData.c as Record<string, unknown>[]).map((e) => ({
        id: e.a, year: e.b, configuredHours: e.c, modifiedAt: e.d, syncedAt: e.e, isDeleted: e.f,
      }))
    : [];

  const shifts = Array.isArray(legacyData.d)
    ? (legacyData.d as Record<string, unknown>[]).map((e) => ({
        id: e.a, name: e.b, icon: e.c, backgroundColor: e.d, startTime: e.e,
        endTime: e.f, hoursWorked: e.g, isActive: e.h, createdAt: e.i,
        modifiedAt: e.j, syncedAt: e.k, isDeleted: e.l,
      }))
    : [];

  const reminders = Array.isArray(legacyData.e)
    ? (legacyData.e as Record<string, unknown>[]).map((e) => ({
        id: e.a, name: e.b, icon: e.c, backgroundColor: e.d, isActive: e.e,
        seriesFrequency: 'never', seriesEndDate: null,
        createdAt: e.f, modifiedAt: e.g, syncedAt: e.h, isDeleted: e.i,
      }))
    : [];

  const syncConfig = Array.isArray(legacyData.f)
    ? (legacyData.f as Record<string, unknown>[])
    : [];

  return {
    metadata: {
      createdAt: legacyMeta.a ?? '',
      appVersion: '1.0.0',
      platform: 'web',
      schemaVersion: 1,
    },
    data: { calendarEvents, notificationRecords, annualHoursConfig, shifts, reminders, syncConfig },
  };
};

/**
 * Deserializes a JSON string into a BackupFile structure.
 *
 * Supports both the current format (named keys: metadata, data, id, eventType, etc.)
 * and the legacy minified format (single-letter keys: a, b, c, etc.) produced by
 * earlier versions of the app.
 *
 * - Parses the JSON and extracts only known fields (unknown fields are ignored
 *   at any nesting level — forward compatibility per Requirement 9.6, 10.4).
 * - All date fields remain as ISO 8601 strings (the BackupFile interfaces use
 *   string types, not Date objects).
 * - Preserves UUIDs as-is and handles null fields correctly.
 */
export const deserializeBackup = (json: string): BackupFile => {
  let parsed = JSON.parse(json) as Record<string, unknown>;

  // Detect and convert legacy minified format
  if (isLegacyFormat(parsed)) {
    parsed = convertLegacyFormat(parsed);
  }

  const rawMetadata = (parsed.metadata ?? {}) as Record<string, unknown>;
  const rawData = (parsed.data ?? {}) as Record<string, unknown>;

  const metadata: BackupMetadata = {
    createdAt: String(rawMetadata.createdAt ?? ''),
    appVersion: String(rawMetadata.appVersion ?? ''),
    platform: rawMetadata.platform as BackupMetadata['platform'],
    schemaVersion: Number(rawMetadata.schemaVersion ?? 0),
  };

  const data: BackupData = {
    calendarEvents: mapArray(
      rawData.calendarEvents,
      mapCalendarEvent,
    ),
    notificationRecords: mapArray(
      rawData.notificationRecords,
      mapNotificationRecord,
    ),
    annualHoursConfig: mapArray(
      rawData.annualHoursConfig,
      mapAnnualHoursConfig,
    ),
    shifts: mapArray(rawData.shifts, mapShift),
    reminders: mapArray(rawData.reminders, mapReminder),
    syncConfig: mapArray(rawData.syncConfig, mapSyncConfig),
  };

  return { metadata, data };
};

const mapArray = <T>(
  raw: unknown,
  mapper: (item: Record<string, unknown>) => T,
): T[] => {
  if (!Array.isArray(raw)) return [];
  return raw.map((item) => mapper(item as Record<string, unknown>));
};

const mapShift = (raw: Record<string, unknown>): BackupShift => ({
  id: String(raw.id ?? ''),
  name: String(raw.name ?? ''),
  icon: String(raw.icon ?? ''),
  backgroundColor: String(raw.backgroundColor ?? ''),
  startTime: Number(raw.startTime ?? 0),
  endTime: Number(raw.endTime ?? 0),
  hoursWorked: Number(raw.hoursWorked ?? 0),
  isActive: Boolean(raw.isActive),
  createdAt: String(raw.createdAt ?? ''),
  modifiedAt: String(raw.modifiedAt ?? ''),
  syncedAt: raw.syncedAt === null || raw.syncedAt === undefined
    ? null
    : String(raw.syncedAt),
  isDeleted: Boolean(raw.isDeleted),
});

const mapReminder = (raw: Record<string, unknown>): BackupReminder => ({
  id: String(raw.id ?? ''),
  name: String(raw.name ?? ''),
  icon: String(raw.icon ?? ''),
  backgroundColor: String(raw.backgroundColor ?? ''),
  isActive: Boolean(raw.isActive),
  seriesFrequency: String(raw.seriesFrequency ?? 'never'),
  seriesEndDate: raw.seriesEndDate === null || raw.seriesEndDate === undefined
    ? null
    : String(raw.seriesEndDate),
  createdAt: String(raw.createdAt ?? ''),
  modifiedAt: String(raw.modifiedAt ?? ''),
  syncedAt: raw.syncedAt === null || raw.syncedAt === undefined
    ? null
    : String(raw.syncedAt),
  isDeleted: Boolean(raw.isDeleted),
});

const mapCalendarEvent = (
  raw: Record<string, unknown>,
): BackupCalendarEvent => ({
  id: String(raw.id ?? ''),
  eventType: String(raw.eventType ?? ''),
  eventTypeId: String(raw.eventTypeId ?? ''),
  startDay: String(raw.startDay ?? ''),
  endDay: String(raw.endDay ?? ''),
  startTime: Number(raw.startTime ?? 0),
  endTime: Number(raw.endTime ?? 0),
  totalHours: Number(raw.totalHours ?? 0),
  notes: raw.notes === null || raw.notes === undefined
    ? null
    : String(raw.notes),
  alertOffsets: Array.isArray(raw.alertOffsets)
    ? (raw.alertOffsets as unknown[]).map(Number)
    : [],
  seriesId: raw.seriesId === null || raw.seriesId === undefined
    ? null
    : String(raw.seriesId),
  modifiedAt: String(raw.modifiedAt ?? ''),
  syncedAt: raw.syncedAt === null || raw.syncedAt === undefined
    ? null
    : String(raw.syncedAt),
  isDeleted: Boolean(raw.isDeleted),
});

const mapNotificationRecord = (
  raw: Record<string, unknown>,
): BackupNotificationRecord => ({
  id: String(raw.id ?? ''),
  calendarEventId: String(raw.calendarEventId ?? ''),
  alertOffset: Number(raw.alertOffset ?? 0),
  triggerTime: String(raw.triggerTime ?? ''),
  isDelivered: Boolean(raw.isDelivered),
  isRead: Boolean(raw.isRead),
  modifiedAt: String(raw.modifiedAt ?? ''),
  syncedAt: raw.syncedAt === null || raw.syncedAt === undefined
    ? null
    : String(raw.syncedAt),
  isDeleted: Boolean(raw.isDeleted),
});

const mapAnnualHoursConfig = (
  raw: Record<string, unknown>,
): BackupAnnualHoursConfig => ({
  id: String(raw.id ?? ''),
  year: Number(raw.year ?? 0),
  configuredHours: Number(raw.configuredHours ?? 0),
  modifiedAt: String(raw.modifiedAt ?? ''),
  syncedAt: raw.syncedAt === null || raw.syncedAt === undefined
    ? null
    : String(raw.syncedAt),
  isDeleted: Boolean(raw.isDeleted),
});

const mapSyncConfig = (raw: Record<string, unknown>): BackupSyncConfig => ({
  serverUrl: String(raw.serverUrl ?? ''),
  apiKey: String(raw.apiKey ?? ''),
  username: String(raw.username ?? ''),
  apiBasePath: String(raw.apiBasePath ?? ''),
  syncIntervalMinutes: Number(raw.syncIntervalMinutes ?? 0),
  isPaused: Boolean(raw.isPaused),
  lastSyncedAt: raw.lastSyncedAt === null || raw.lastSyncedAt === undefined
    ? null
    : String(raw.lastSyncedAt),
});
